cmd/update.go

```diff
@@ -0,0 +1,496 @@
+	url := githubAPIBase + "/repos/oasisprotocol/cli/releases/latest"
+
+	for {
+		req, _ := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+		req.Header.Set("User-Agent", "oasis-cli-updater")
+		req.Header.Set("Accept", "application/vnd.github+json")
+		req.Header.Set("X-GitHub-Api-Version", "2022-11-28")
+		if t := os.Getenv("GITHUB_TOKEN"); t != "" {
+			req.Header.Set("Authorization", "Bearer "+t)
+		}
+
+		res, err := httpClient.Do(req)
+		if err != nil {
+			return nil, err
+		}
+
+		// Respect GitHub throttling hints.
+		if res.StatusCode == http.StatusForbidden {
+			if sec, _ := strconv.Atoi(res.Header.Get("Retry-After")); sec > 0 && sec < 60 {
+				res.Body.Close() // close before sleeping to free the connection
+				time.Sleep(time.Duration(sec) * time.Second)
+				continue
+			}
+		}
+
+		if res.StatusCode != http.StatusOK {
+			err = fmt.Errorf("GitHub API: %s", res.Status)
+			res.Body.Close()
+			return nil, err
+		}
+
+		var rel ghRelease
+		decodeErr := json.NewDecoder(res.Body).Decode(&rel)
+		res.Body.Close()
+		return &rel, decodeErr
+	}
+}
+
+// Asset selection
+
+var archRE = map[string]*regexp.Regexp{
+	"amd64": regexp.MustCompile(`(?:^|[-_])(amd64|x86_64)(?:[-_.]|$)`),
+	"arm64": regexp.MustCompile(`(?:^|[-_])(arm64|aarch64|universal)(?:[-_.]|$)`),
+}
+
+// archAliases returns candidate substrings ordered by preference.
+// The "universal" token is intentionally included to match Apple's
+// multi-arch binaries (e.g. "darwin_arm64_universal") and is harmless on
+// other platforms where the substring may also appear.
+func archAliases(arch string) []string {
+	switch arch {
+	case "amd64":
+		return []string{"amd64", "x86_64", "universal", "all"}
+	case "arm64":
+		return []string{"arm64", "aarch64", "universal", "all"}
+	default:
+		return []string{arch, "all"}
+	}
+}
+
+func matches(goos, goarch string, asset string) bool {
+	if !strings.Contains(asset, goos) {
+		return false
+	}
+	if re, ok := archRE[goarch]; ok {
+		return re.MatchString(asset)
+	}
+	return strings.Contains(asset, goarch)
+}
+
+func findAssetFor(rel *ghRelease, goos, goarch string) (name, url string, err error) {
+	for _, wantArch := range archAliases(goarch) {
+		for _, a := range rel.Assets {
+			if matches(goos, wantArch, strings.ToLower(a.Name)) {
+				return a.Name, a.BrowserDownloadURL, nil
+			}
+		}
+	}
+	names := make([]string, 0, len(rel.Assets))
+	for _, a := range rel.Assets {
+		names = append(names, a.Name)
+	}
+	return "", "", fmt.Errorf("no asset for %s/%s (available: %v)", goos, goarch, names)
+}
+
+func findAsset(rel *ghRelease) (name, url string, err error) {
+	return findAssetFor(rel, runtime.GOOS, runtime.GOARCH)
+}
+
+func findSha256For(rel *ghRelease, binName string) (url string, ok bool) {
+	want := binName + ".sha256"
+	for _, a := range rel.Assets {
+		if a.Name == want {
+			return a.BrowserDownloadURL, true
+		}
+	}
+	return "", false
+}
+
+// Download with progress
+
+type progressWriter struct {
```

Comment:
Move the progress bar logic to a common place. It will be used by https://github.com/oasisprotocol/cli/issues/522 too.